client/connection.go

```diff
@@ -185,16 +186,19 @@ func ConnectLXDUnixWithContext(ctx context.Context, path string, args *Connectio
 		eventListeners:     make(map[string][]*EventListener),
 	}
 
-	// Determine the socket path
+	// Determine the socket path.
 	if path == "" {
 		path = os.Getenv("LXD_SOCKET")
 		if path == "" {
 			lxdDir := os.Getenv("LXD_DIR")
-			if lxdDir == "" {
-				lxdDir = "/var/lib/lxd"
+			if lxdDir != "" {
+				path = filepath.Join(lxdDir, "unix.socket")
+			} else {
+				path = "/var/lib/lxd/unix.socket"
+				if !shared.PathIsWritable(shared.HostPath(path)) {
+					path = "/var/snap/lxd/common/lxd/unix.socket"
```

Comment:
Is this change required? How is the `lxc` command working inside the snap, does it set LXD_DIR env?

Are you adding this for lxd-migrate's benefit?

Reply:
I think we should re-arrange it so that if /var/snap/lxd/common/lxd/unix.socket is writable we use it, otherwise fallback to /var/lib/lxd/unix.socket

Reply:
> How is the lxc command working inside the snap, does it set LXD_DIR env?

Yes.

https://github.com/canonical/lxd-pkg-snap/blob/latest-edge/snapcraft/commands/lxc#L32-L38

> Is this change required?

The `lxd-migrate` is not snapped, so it will not have `LXD_DIR` set. Therefore, we can avoid this change and make `lxd-migrate` check for the snapped LXD socket explicitly. 

However, I wanted to take this opportunity to change this, as it will benefit `lxd-migrate`, Terraform provider, and probably pylxd (?) - which are all currently *trying* which socket exists as they are not snapped.

> I think we should re-arrange it so that if /var/snap/lxd/common/lxd/unix.socket is writable we use it, otherwise fallback to /var/lib/lxd/unix.socket

Fixed.